### DISCUSSION TITLE
Tabs and windows

### DISCUSSION THREAD

**Discussion by MRoth1910** [musescore/MuseScore#15882] [Ideas]
(If a new conversation was started pursuant to discussion in https://github.com/musescore/MuseScore/issues/12647, please move this!)

I'm a novice MS user and am still learning about macOS, and I don't have a programming background, so please bear with me.

I won't reiterate what's been said about how MS4 works on macOS. I would, however, like to say that leaving tab decisions to macOS would not be a great solution if I am understanding what has been said so far (and I can appreciate that existing code may have been, well, difficult to reimplement). 

Being able to open a new window, then a new tab as needed, or a new window for a separate project while keeping another window open is a good thing, and I would prefer if the function came from within the program or if it was able to cleanly merge two windows (e.g. [what's possible in Safari and in Preview](https://apple.stackexchange.com/questions/392285/how-do-i-merge-the-tabs-of-two-open-preview-windows-if-i-have-more-than-two-open)) and was then cleanly documented in the handbook.

**Comment by cbjeukendrup** [comment]
As said, the problem is currently that every score is opened not only in a new _window_, but even in a new _process_. This results in multiple dock icons and other bugs, and is really not the way to do things on macOS. The reason that we still do it that way, is that there were technical difficulties in supporting multiple scores at the same time in the same process, given MS4's much more advanced audio engine. So first we need to fix that, and do everything in one process; only then, we can start looking into the question "tabs or windows".

Once we get everything in one process again, most likely we'll initially go back to the MS3 approach, so everything in one window with multiple tabs. This is _much_ easier to support than multiple windows in one process.

When we do get the chance to support multiple windows in one process as a third step, then as a fourth step we can look into allowing macOS to let you turn windows in tabs. However, I doubt whether that will work; probably it will look ugly and work badly. One problem is that we then have two tab bars: one inside the MuseScore window (for switching between parts), and one at the very top of the window created by macOS (that cannot be customised). Also, we'll need to build a custom tab system anyway, because Windows and Linux. 

Someone managed to implement a macOS-like system using a custom tab bar. Maybe we can use that... 

Anyway, all of that is not trivial, probably not realistic in the short term.

**Comment by MRoth1910** [comment]
Yeah. I got the first part and really wasn't responding to that. ¯\_(ツ)_/¯ I've complained elsewhere that the multiple processes part isn't good.

I just don't want the tabs "left to the OS" because that result is horrible and inconsistent, nor is it intuitive.

It is understandable but extremely disappointing to hear that something included as a design wound up being essentially a bug; the primary public faces of the development didn't seem to understand that a) regardless of the technical aspects unseen to most users (although noticeable if you use the new Stage Manager feature) the tabs are important and b) that this is really bad for macOS users. I'm not trying to beat a dead horse, and I'm making the most of the new version, but to hear that this is going to require a very long wait for a fix is extremely disappointing.
